File: gs/src/gate/Handler.java
package gate;

import common.IHandler;
import msg.server2gate.Gate2XDelSession;
import msg.server2gate.Gate2XForwardFromClientSession;
import msg.server2gate.X2XAnnounceServer;

/**
 * Created by HuangQiang on 2017/6/1.
 */
public enum Handler implements IHandler {
    Ins;

    public void bind() {
        X2XAnnounceServer.handler = this::process;
        Gate2XDelSession.handler = this::process;
        Gate2XForwardFromClientSession.handler = this::process;
    }

    public void process(X2XAnnounceServer msg) {
        log.debug("== process {}", msg);
    }

    public void process(Gate2XDelSession msg) {
        log.debug("== process {}", msg);
    }

    public void process(Gate2XForwardFromClientSession msg) {
        GateClient.getIns().process(msg);
    }
}

File: gs/src/game/auth/Handler.kt
package game.auth

import annotations.Handler
import common.IHandler
import common.session
import gate.GateClient
import msg.gs.auth.CAuth
import perfect.common.Trace.log
import perfect.db.Database
import perfect.txn.Transaction

/**
 * Created by HuangQiang on 2017/6/5.
 */
@Handler
object Handler : IHandler {
    override fun bind() {
        CAuth.handler = CAuth.Handler { PCAuth(it).execute() }
    }

   private class PCAuth(val m : CAuth) : perfect.txn.Procedure() {
       override fun process(): Boolean {

           if(m.account.isEmpty()) {
               log.debug("CAuth. account is empty")
               return false
           }
           var account = db.auth.Accounts.get(m.account)
           if(account == null) {
               account = db.auth.Account.newBean()
               val now = System.currentTimeMillis()
               account.createtime = now
               db.auth.Accounts.insert(m.account, account)
               log.info("create account:{}", m.account)
           }
           Transaction.syncExecuteWhileCommit({
               GateClient.getIns().auth(m.session, m.account)
           })
           return true
       }
   }
}